ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+  if (err != RC_OK) {
+    lock_handle_unlock(&ps->lock_handle);
+    log_warning(logger_id, "Local snapshots pruning has failed with error code: %d\n", err);
+  }
+
+  return RC_OK;
+}
+
+retcode_t iota_local_snapshots_pruning_service_init(pruning_service_t *const ps,
+                                                    snapshots_provider_t *const snapshot_provider,
+                                                    spent_addresses_service_t *const spent_addresses_service,
+                                                    tips_cache_t *const tips_cache,
+                                                    iota_consensus_conf_t const *const conf) {
+  logger_id = logger_helper_enable(PRUNING_SERVICE_LOGGER_ID, LOGGER_DEBUG, true);
+  memset(ps, 0, sizeof(pruning_service_t));
+  ps->conf = conf;
+  ps->last_pruned_snapshot_index = snapshot_provider->inital_snapshot.metadata.index;
+  ps->last_snapshot_index_to_prune = ps->last_pruned_snapshot_index;
+  ps->spent_addresses_service = spent_addresses_service;
+  ps->tips_cache = tips_cache;
+  lock_handle_init(&ps->lock_handle);
+  return RC_OK;
+}
+
+retcode_t iota_local_snapshots_pruning_service_start(pruning_service_t *const ps) {
+  if (ps == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  ps->running = true;
+
+  log_info(logger_id, "Spawning pruning service thread\n");
+  if (thread_handle_create(&ps->pruning_service_thread, (thread_routine_t)pruning_service_routine, ps) != 0) {
+    log_critical(logger_id, "Spawning pruning service thread failed\n");
+    return RC_THREAD_CREATE;
+  }
+
+  return RC_OK;
+}
+
+retcode_t iota_local_snapshots_pruning_service_stop(pruning_service_t *const ps) {
+  retcode_t ret = RC_OK;
+
+  if (ps == NULL) {
+    return RC_NULL_PARAM;
+  } else if (ps->running == false) {
+    return RC_OK;
+  }
+
+  ps->running = false;
+  cond_handle_signal(&ps->cond_pruning_service);
+  log_info(logger_id, "Shutting down pruning service thread\n");
+  if (thread_handle_join(ps->pruning_service_thread, NULL) != 0) {
+    log_error(logger_id, "Shutting down pruning service thread failed\n");
+    ret = RC_THREAD_JOIN;
+  }
+
+  return ret;
+}
+
+retcode_t iota_local_snapshots_pruning_service_destroy(pruning_service_t *const ps) {
+  retcode_t ret = RC_OK;
+
+  if (ps == NULL) {
+    return RC_NULL_PARAM;
+  } else if (ps->running) {
+    return RC_STILL_RUNNING;
+  }
+
+  cond_handle_destroy(&ps->cond_pruning_service);
+  memset(ps, 0, sizeof(pruning_service_t));
+  logger_helper_release(logger_id);
+  lock_handle_destroy(&ps->lock_handle);
+
+  return ret;
+}
+
+void iota_local_snapshots_pruning_service_update_current_snapshot(pruning_service_t *const ps,
+                                                                  snapshot_t *const snapshot) {
+  lock_handle_lock(&ps->lock_handle);
+  ps->last_snapshot_index_to_prune = snapshot->metadata.index;
+  lock_handle_unlock(&ps->lock_handle);
+
+  ps->new_snapshot = snapshot;
+  cond_handle_signal(&ps->cond_pruning_service);
```

Comment:
Why is there a timer in the first place since we can signal it when we need it ? I don't think we need a timer at all.

Reply:
because we want to wake up between `snapshot_index` to `snapshot_index + 1` even when we are not signaled to allow updating `last_snapshot_index_to_prune`